### PR TITLE
Function Contravariance

### PR DESCRIPTION
**Pull request by RGates94**
For soundness it is necessary for function arguments and return types to be contravariant, this PR makes subtyping behave in this way.

### CODE REVIEW

File: src/compile/ast.rs [hunk 1]
```diff
@@ -194,8 +194,8 @@ impl Type {
             Type::Func(is_impure, args, ret) => {
                 if let Type::Func(is_impure2, args2, ret2) = rhs {
                     (*is_impure || !is_impure2)
-                        && arg_vectors_assignable(args, args2, type_tree, seen.clone())
-                        && (ret2.assignable(ret, type_tree, seen)) // note: rets in reverse order
+                        && arg_vectors_assignable(args2, args, type_tree, seen.clone())
+                        && (ret.assignable(ret2, type_tree, seen)) // note: rets in reverse order
```
Comment (edfelten) [inline src/compile/ast.rs:198]: Should the comment "note: rets in reverse order" be revised?

**Review by edfelten — COMMENTED** [review]
One question to resolve before approving this.

File: src/compile/ast.rs (hunk 1 above)
Comment (RGates94) [reply src/compile/ast.rs:198]: Yes, I'll edit this.

**Review by edfelten — APPROVED** [review]
Looks good.

**Pull request merged**